README.md
> ### 项目架构

项目基于python+flask+mysql+gunicorn+supervisor进行搭建的一个web脚手架，架子已经基本搭建起来，备具Restful API、验证session、页面模板等功能

git clone之后修改配置即可运行，在此基础上可进行二次开发，可以前端/后台独立、也可以运用flask的jinja2模板

  - python 开发语言，基于python2
  - flask 使用的web框架
  - mysql 数据库
  - gunicorn web服务与进程
  - supervisor 项目启动、停止、重启等操作

项目主要运行于Linux系统上，本人使用Centos7.0

> ### 配置说明

项目配置主要有2套，位于项目的根目录etc下
  - dev 测试环境
  - prod 线上环境

每套配置文件夹下有3个配置文件，config与gunicorn进行绑定：
  - config：项目的db、mail、log等配置，这里的log记录项目的log，关于项目相关的配置都可以在此文件进行配置
  - gunicorn：项目启动时所需要的IP、port、log、进程数量等配置
  - supervisor: 项目管理的配置信息

.yaml格式的配置文件是有deploy/config.py进行解析的，如果在config.yaml配置文件中添加配置信息，需要在此文件进行解析添加，建议添加配置默认值

> ### 环境搭建

  - Centos7.0系统服务器
  - mysql数据库安装与配置
  - git clone https://github.com/GIS90/base_webframe.git
  - 更新web配置文件：etc/prod/config.py（线上）、etc/dev/config.py（测试），根据不同需求进行项目更改
  - 安装项目运行的环境：python install_env.py，建立项目独立的运行环境，安装了virtualenv、python、gunicorn、packages等操作，了解具体详情请参考代码
  - 项目根目录：source .venv/bin/activate：激活项目环境
  - 启动项目：gunicorn -c etc/prod/gunicorn.conf（线上）、gunicorn -c etc/dev/gunicorn.conf（测试）或者手动启动（下面有介绍），关于dev/prod配置请参考上面介绍
  - 选做：安装supervisor && 项目加入supervisor进行管理，项目包含了supervisord配置文件&&项目supervisor配置文件

> ### 安装包

  - pip install -r requirements.txt

此程序运行于python2，其中requirements.txt项目所需要的包，已固定版本

> ### supervisor

管理项目进程的启动、停止、重启等操作
安装：pip install supervisor
配置：
  - dev：etc/dev/supervisor_base_webframe.conf
  - prod：etc/prod/supervisor_base_webframe.conf

把指定环境的supervisor_base_webframe.conf cp到/etc/supervisord.d/include/*下
项目root根目录下有supervisord.conf文件，用来配置supervisord

> ### 管理gunicorn

负责web项目进程、服务

安装：pip install gunicorn

配置：
  - dev：etc/dev/gunicorn.conf
  - prod：etc/prod/gunicorn.conf

如需特别项目启动信息，可以加入gunicorn.conf或者更改命令行gunicorn启动方式加入参数即可

> ### 其他

  - base_webframe_start.bash、base_webframe_end.bash为手动方式进行项目启动与项目结束（已废弃）
  - deploy>utils>utils.py 为工具方法，任何python2项目都适合使用

> ### sql

创建sql：deploy>sql>table_create_sql.sql，直接执行即可

插入sql：deploy>sql>print_insert_sql.py，具体需要哪些insert的数据，看代码，根据需要插入

> ### 手动启动

项目根目录wsgi.py文件加入

app.run(host="0.0.0.0", port=11111, debug=True)

执行sudo python wsgi.py，代码目前已写入，处于注释状态

通过手动启动的项目为dev开发环境配置，可在deploy/config.py中进行默认调整（mode = os.environ.get('mode') or 'dev'）

如果手动启动模式开启，在gunicorn进行启动，会error: [Errno 48] Address already in use.

注意：启动项目一定要用virtualenv安装的python环境进行启动（source .venv/bin/activate）

> ### 联系方式

* ***Github:*** https://github.com/GIS90

* ***Email:*** gaoming971366@163.com

* ***Blog:*** http://pygo2.cn

* ***WeChat:*** PyGo90


Enjoy the good lift everyday！欢迎大家跟我一起来完善这个脚手架！！


ChangeLog
CHANGES
=======

* fix bug: setter: submit info api
* add logo image
* del avatar image
* optimize: setter.user
* add comment
* start dev branch
* fix bug: run
* install\_env.py:fix virtualenv version
* fix bug: install\_env.py
* add upload image
* complete html sub\_template
* add html sub\_template
* update README.md
* fix bug: update home page employee count
* add edit and softdel employee
* add\_employee add tip
* bank\_id add number rule
* add edit func
* update README.md
* employee、enums更新&&开发信息管理->新增功能
* user update sysuser
* blueprint: tasks update apis
* add 404&500 page
* app.context\_processor: menu&&current\_user
* set favicon
* important file add note
* README.md update
* add dev/prod config.yaml
* console\_scripts update
* add AppBaseClass && add setup.cfg console\_scripts
* update README.md
* delete backup setup.py config
* first commit


deploy/__init__.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

"""
--------------------------------------------------------------
describe:
    WebFlaskServer initialize 
    flask app and blueprint collections
    
    Singleton mode

base_info:
    __version__ = "v.10"
    __author__ = "mingliang.gao"
    __time__ = "2020/5/19 3:42 PM"
    __mail__ = "mingliang.gao@163.com"
--------------------------------------------------------------
"""

# ------------------------------------------------------------
# usage: /usr/bin/python __init__.py.py
# ------------------------------------------------------------
import os
import sys
from flask import (Flask,
                   render_template,
                   session,
                   redirect,
                   url_for,
                   g,
                   request)

from deploy.config import VERSION, NAME, SECRET_KEY
from deploy.models.base import get_session
from deploy.utils.base_class import WebBaseClass
from deploy.utils.logger import logger as LOG
from deploy.utils.utils import get_user_id
from deploy.views.employee import employee
from deploy.views.apis import apis
from deploy.views.manage import manage
from deploy.views.home import home
from deploy.views.setter import setter
from deploy.services.sysuser import SysUserService


app = Flask(__name__)


class WebFlaskServer(WebBaseClass):
    app = None
    version = VERSION
    name = NAME

    def __init__(self, app):
        """
        Initialize webFlaskServer instance
        and flask configuration
        """
        self.app = app
        if not self.app:
            LOG.info('Web server initialize is failure......')
            sys.exit(1)

        _realpath = os.path.dirname(os.path.realpath(__file__))
        self.app.template_folder = _realpath + '/templates/'
        self.app.secret_key = SECRET_KEY or 'python'
        self.app.static_folder = _realpath + '/static'
        self.app.static_url_path = '/static'
        self.app.add_url_rule(self.app.static_url_path + '/<path:filename>',
                              endpoint='static',
                              view_func=self.app.send_static_file)

        super(WebFlaskServer, self).__init__()

        @self.app.before_request
        def before_request():
            if get_user_id():
                return
            # api: rest apis
            # manage: login apis
            if request.blueprint in ['api', 'manage', None]:
                return
            # special api for blueprints
            if request.endpoint.endswith('ForApi') or \
                    request.endpoint.endswith('for_api'):
                return

            return redirect(url_for('manage.index'))

        @self.app.before_first_request
        def before_first_request():
            g._session = get_session()

        @self.app.errorhandler(404)
        def not_found_error(error):
            LOG.error("%s is not found 404" % request.url)
            return render_template('errors/404.html', ), 404

        @self.app.errorhandler(500)
        def server_error(error):
            LOG.error("%s is server error 500" % request.url)
            return render_template('errors/500.html'), 500

        @self.app.context_processor
        def default_context_processor():
            user_id = session.get('user_id')
            menu = current_user = dict()
            if hasattr(g, 'menuf'):
                menu['f'] = g.menuf or 'index'
            if hasattr(g, 'menusub'):
                menu['sub'] = g.menusub or 'index'
            if user_id:
                current_user = SysUserService().get_user_by_params(user_id)

            return {
                'current_user': current_user,
                'sysversion': VERSION,
                'menu': menu
            }

        # set favicon
        @self.app.route('/favicon.ico')
        def get_defaule_favicon():
            return self.app.send_static_file('images/favicon.ico')

    def register_blueprint(self, obj_n, obj):
        """
        view blueprint register
        :param obj_n: blueprint object
        :param obj: blueprint name
        :return: None
        """
        if obj:
            LOG.info('Blueprint %s is register' % obj_n)
            self.app.register_blueprint(obj)

    def _autoinit_register_blueprint(self):
        self.register_blueprint('employee', employee)
        self.register_blueprint('apis', apis)
        self.register_blueprint('manage', manage)
        self.register_blueprint('home', home)
        self.register_blueprint('setter', setter)

    def init_run(self):
        LOG.debug('Server is initializing......')
        self._autoinit_register_blueprint()
        LOG.info('Web server is running......')


def create_app():
    return WebFlaskServer(app).app
